Make a workspace run without any user env

garagatyi opened the pull request:
### What does this PR do?
This PR allows running Che 7 workspace without any user environment but with Che 7 tooling set.
Workspace without environment has an empty list of environments and null value in `defaultEnv` field.
Field `activeEnv` is supposed to be `null` too.
Video: https://youtu.be/QeeRtoYZU-o

**Why**
It is useful to be able to run Theia workspace for Theia plugins development without any additional machines since Theia has hosted mode.
Another use-case is when IDE container has everything needs for a user to run his application - great for starting a new project that is lightweight and doesn't have tons of apps in the deployment.

**What is changed:**
- Migration of DB:
 - make `defaultEnv` nullable
 - remove env_name from runtimes primary key
 - make `env_name` in runtimes nullable
- Make code respect the fact that there are workspaces with no env
- settings API that returns the list of supported environment types returns `no-environment` type that indicates that infrastructure support workspace with no environment at all.  


Checked:
- Che 7 on OS infra
- Che 7 on OS infra with GWT IDE sidecar
- Che 7 with Postgres and Auth
- Che 7 with Postgres and existing+running+starting workspaces  (rolling update)
- Che 7 on k8s infra

To check:
- RhChe 7

Doesn't support:
- Docker 
- Che 6 on all the infras
- rolling update with H2 (not supported in Che)

Todo:
- fix UD
- fix factory acceptance

### What issues does this PR fix or reference?

<!-- #### Changelog -->
<!-- The changelog will be pulled from the PR's title. 
     Please provide a clear and meaningful title to the PR and don't include issue number -->


#### Release Notes
<!-- markdown to be included in marketing announcement - N/A for bugs -->


#### Docs PR
<!-- Please add a matching PR to [the docs repo](https://github.com/eclipse/che-docs) and link that PR to this issue.
Both will be merged at the same time. -->

garagatyi commented:
I added a demo of how it works in the PR description

skabashnyuk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/model/KubernetesRuntimeState.java:
```diff
@@ -13,26 +13,31 @@
 @NamedQueries({
-  @NamedQuery(name = "KubernetesRuntime.getAll", query = "SELECT r FROM KubernetesRuntime r"),
-  @NamedQuery(
-      name = "KubernetesRuntime.getByWorkspaceId",
```
@sleshchenko wdyt?

sleshchenko replied:
@skabashnyuk What is a question? It's OK for me to remove this named query since workspace id became the primary key.

skabashnyuk replied:
yes

garagatyi replied:
@sleshchenko for now workspaceId+ownerId is the primary key. But I believe it doesn't change what you wanted to say, right?

sleshchenko replied:
@garagatyi Thanks for the clarification. Right, it doesn't change. I meant JPA ID (javax.persistence.Id) but not Database primary key.

garagatyi commented on wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.14.0/1__remove_not_null_constraint_from_field.sql:
```diff
@@ -0,0 +1,21 @@
+--
+
+-- Allow null as defaultEnv field in workspace config
+ALTER TABLE workspaceconfig ALTER COLUMN defaultenv VARCHAR(255) NULL;
```
ok, I'll take a look at how to name it better.

garagatyi commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java:
```diff
@@ -782,17 +795,22 @@ private boolean containsThisRuntimesId(String workspaceId) {
 
   public Set<String> getSupportedRecipes() {
-    return environmentFactories.keySet();
+    return Sets.union(environmentFactories.keySet(), Collections.singleton("no-environment"));
```
You are right. I forgot to do this

l0rd replied:
@sleshchenko what's an example of infrastructure that doesn't support `no-environment`?

sleshchenko replied:
For now, it's only docker infrastructure. I prefer to make approach flexible and allow not support running `no-environments` if it is possible.

l0rd replied:
Ok I see. But that's only temporary. We are evaluating if Docker infrastructure will be supported or not in Che 7 but, if it will be supported, we have to implement `no-environment` for it as well. In other words we should not complicate the implementation to allow infrastructures that do not support `no-environment` (because all should support it).

l0rd commented:
This is cool stuff @garagatyi. I like the demo. A few comments:

> **Why**
It is useful to be able to run Theia workspace for Theia plugins development without any additional machines since Theia has hosted mode.

Just to make sure we are on the same page this is not only for hosted Theia scenario. This is for all the cases where a developer doesn't use a target runtime. That may be because he is in the early stage of the development or he doesn't want to load the runtime in the IDE to have a lighter environment. And he can avoid having a runtime because the IDE tools have already everything he needs to develop/test/debug.

> Che 7 with GWT IDE sidecar

This is a nice to have of course. But if some changes are needed to the GWT IDE sidecar these should be out of the scope of this PR (i.e. priorities are different).

garagatyi commented:
@l0rd thanks for mentioning, there are several use-cases for that.
I'll take into account your comment about GWT IDE

garagatyi commented:
@skabashnyuk @sleshchenko @mshaposhnik this PR is not in WIP anymore

garagatyi commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java:
```diff
@@ -785,10 +793,16 @@ private boolean containsThisRuntimesId(String workspaceId) {
-    String recipeType = environment.getRecipe().getType();
+    String recipeType;
+    if (environment == null) {
+      recipeType = Constants.NO_ENVIRONMENT_RECIPE_TYPE;
```
I like the idea but we use `no-environment` recipe returned in the list of supported environments to allow clients (aka UD) to use it properly. Because of that, keeping it as one of the environment implementations simplifies the code in my opinion. I would leave it like that for the time being.

sleshchenko replied:
Good point with UD, I missed it before. I'm OK with the current approach since I can't propose an easy alternative way which would expose supporting of `no-environment` for clients.

garagatyi commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java:
```diff
@@ -351,24 +351,31 @@ public void stopWorkspace(String workspaceId, Map<String, String> options)
-              workspace.getNamespace(), workspace.getConfig().getName(), envName));
+    // Sidecar-based workspaces allowed not to have any environments
+    String env = null;
+    if (!SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(workspace.getConfig().getAttributes())
```
What about:
```
boolan nonSidecarWorkspace = !SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(workspace.getConfig().getAttributes());
boolean envListNonEmpty = workspace.getConfig().getEnvironments().isEmpty();
if (nonSidecarWorkspace || envName != null || envListNonEmpty) {
// check env
}
```
Which means that body of the if clause would be executed if any of next conditions is true:
- it is Che 6 WS
- default env is set 
- environments list is empty 

WDYT?

sleshchenko replied:
Well, I investigated this code and discovered an interesting thing for me:
`!workspace.getConfig().getEnvironments().isEmpty()` means that if there is one environment then this environment must be configured as default. Am I right?

WorkspaceValidator checks that `DefaultEnv` references existing environment, then maybe it would be better to rewrite check like 
```
if (nonSidecarWorkspace || envName != null || workspace.getConfig().getDefaultEnv) {
}
```

Then it may be rewritten even like in the following way:
```
   if (envName == null) {
      envName = workspace.getConfig().getDefaultEnv()
    }

    // Sidecar-based workspaces allowed not to have any environments
    if (!SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(workspace.getConfig().getAttributes())
        || envName != null) {
      if (envName == null) {
          throw new NotFoundException("Missing environment to start");
      }
      if (!workspace.getConfig().getEnvironments().containsKey(envName)) {
        throw new NotFoundException(
            format(
                "Workspace '%s:%s' doesn't contain environment '%s'",
                workspace.getNamespace(), workspace.getConfig().getName(), envName));
      }

      // validate environment in advance
      try {
        runtimes.validate(workspace.getConfig().getEnvironments().get(envName));
      } catch (InfrastructureException | ValidationException e) {
        throw new ServerException(e);
      }
    }
```

sleshchenko replied:
More more way:
```java
  /** Asynchronously starts given workspace. */
  private void startAsync(
      WorkspaceImpl workspace, @Nullable String envName, Map<String, String> options)
      throws ConflictException, NotFoundException, ServerException {
    String env = getValidEnvironmentName(workspace, envName);
    workspace.getAttributes().put(UPDATED_ATTRIBUTE_NAME, Long.toString(currentTimeMillis()));
    workspaceDao.update(workspace);
    runtimes
        .startAsync(workspace, env, firstNonNull(options, Collections.emptyMap()))
        .thenAccept(aVoid -> handleStartupSuccess(workspace))
        .exceptionally(
            ex -> {
              if (workspace.isTemporary()) {
                removeWorkspaceQuietly(workspace);
              } else {
                handleStartupError(workspace, ex.getCause());
              }
              return null;
            });
  }

  private String getValidEnvironmentName(WorkspaceImpl workspace, String envName)
      throws NotFoundException, ServerException {
      if (envName != null && !workspace.getConfig().getEnvironments().containsKey(envName)) {
        throw new NotFoundException(
            format(
                "Workspace '%s:%s' doesn't contain environment '%s'",
                workspace.getNamespace(), workspace.getConfig().getName(), envName));
      }
    
      envName = firstNonNull(envName, workspace.getConfig().getDefaultEnv());

      if (envName == null && SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(workspace.getConfig().getAttributes()) {
          // Sidecar-based workspaces allowed not to have any environments
          return null; 
      }

      // validate environment in advance
      try {
        runtimes.validate(workspace.getConfig().getEnvironments().get(envName));
      } catch (InfrastructureException | ValidationException e) {
        throw new ServerException(e);
      }
  }
```

garagatyi replied:
> Well, I investigated this code and discovered an interesting thing for me:
!workspace.getConfig().getEnvironments().isEmpty() means that if there is one environment then this environment must be configured as default. Am I right?

absolutely

I like the last approach you suggested, will apply it. Thanks!

sleshchenko commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java:
```diff
@@ -69,23 +68,12 @@ public void validateConfig(WorkspaceConfig config) throws ValidationException, S
-      }
+    // if environments/default environment name are empty and it is sidecar-based workspace
+    // then do not check env since it is a workspace with tooling only
+    if ((config.getEnvironments() != null && !config.getEnvironments().isEmpty())
```
Maybe it would be simpler if you invoke `validateEnvironments(config);` here always, and inside it check
```
  if (environmentIsConfigured && isSidecarWorkspace) {
    do not check env since it is a workspace with tooling only
    return;
  }
```
WDYT?

garagatyi replied:
Yeah, the code might be better in that case. I rework the code a bit to apply this suggestion. 
But it won't that simple because it is OK when both default env and environments are not set.

sleshchenko commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/environment/InternalEnvironmentFactory.java:
```diff
@@ -78,36 +79,40 @@ public InternalEnvironmentFactory(
 
-    machinesValidator.validate(machines);
-
     return doCreate(recipe, machines, warnings);
```
`Recipe` parameter is not marked as `Nullable` but now it is. Please consider adding the corresponding annotations and checks to avoid NPE.

garagatyi replied:
Done

sleshchenko commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java:
```diff
@@ -351,24 +351,31 @@ public void stopWorkspace(String workspaceId, Map<String, String> options)
+
+      // validate environment in advance
+      try {
+        runtimes.validate(workspace.getConfig().getEnvironments().get(env));
```
since defaultEnv is nullable - NPE may happen here

garagatyi commented:
ci-test

garagatyi commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java:
```diff
@@ -69,23 +68,12 @@ public void validateConfig(WorkspaceConfig config) throws ValidationException, S
-      }
+    // if environments/default environment name are empty and it is sidecar-based workspace
+    // then do not check env since it is a workspace with tooling only
+    if ((config.getEnvironments() != null && !config.getEnvironments().isEmpty())
```
Sometimes it is used with DTO objects created by REST framework. As far as I remember it doesn't comply with this policy and does set null collection. 
WDYT?

sleshchenko replied:
As far as I see our DTO implementations comply with this policy https://github.com/eclipse/che/blob/dcb00c53d09fc8ad71b0bc12d44115d7269d2481/core/che-core-api-dto/src/main/java/org/eclipse/che/dto/generator/DtoImplClientTemplate.java#L1346
Example of ServerImplemenation for WorkspaceDto
```java
...
    @Override
    public java.util.Map<java.lang.String, java.lang.String> getAttributes() {
      ensureAttributes();
      return (java.util.Map<java.lang.String, java.lang.String>)(new org.eclipse.che.dto.server.JsonStringMapImpl(attributes));
    }

    protected void ensureAttributes() {
      if (attributes == null) {
        attributes = new java.util.HashMap<java.lang.String, java.lang.String>();
      }
    }
...
```

garagatyi replied:
DTO is not used when Everrest creates those objects.

sleshchenko replied:
Do you mean that in Services we have not to DtoImplementations but objects created by Everrest?

garagatyi replied:
yes,  I believe so

riuvshin commented:
Results of automated E2E tests of Eclipse Che Multiuser on OCP:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1202/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1202//Selenium_20tests_20report/)
docker image: eclipseche/che-server:11890
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

garagatyi commented:
@eclipse/eclipse-che-qa Please, verify tests results

garagatyi commented:
@sleshchenko I think I addressed all the important points. There are some minor things that could go under the radar. Please, take a look.

@skabashnyuk @ibuziuk @mshaposhnik @sleshchenko I changed and improved some stuff including DB migration scripts. Please, take a look or remove yourself from the list of reviewers.

garagatyi commented:
ci-test

skabashnyuk commented:
ok for me in general. I trust @sleshchenko and @mshaposhnik to make a final decision.

sleshchenko commented on wsmaster/integration-tests/cascade-removal/src/test/java/org/eclipse/che/core/db/jpa/CascadeRemovalTest.java:
```diff
@@ -185,7 +185,6 @@ protected void configure() {
                             SshPairImpl.class,
                             VolumeImpl.class,
                             KubernetesRuntimeState.class,
-                            KubernetesRuntimeState.RuntimeId.class,
```
Please remove `RuntimeId` from [persistence.xml](https://github.com/garagatyi/che/blob/noEnvWS/assembly/assembly-wsmaster-war/src/main/resources/META-INF/persistence.xml#L73) as well

garagatyi replied:
Are you sure that this code is up-to-date?

garagatyi replied:
Oh, I see. Thanks!

riuvshin commented:
Results of automated E2E tests of Eclipse Che Multiuser on OCP:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1206/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1206//Selenium_20tests_20report/)
docker image: eclipseche/che-server:11890
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

garagatyi commented:
ci-test

Ohrimenko1988 commented:
It looks like the dependency
```
16:53:33 [INFO] Add the following to your pom to correct the missing dependencies: 
16:53:33 [INFO] 
16:53:33 <dependency>
16:53:33   <groupId>org.eclipse.che.core</groupId>
16:53:33   <artifactId>che-core-commons-annotations</artifactId>
16:53:33   <version>6.15.0-SNAPSHOT</version>
16:53:33 </dependency>
```
should be added to pom.xml

https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1206/console

garagatyi commented:
@Ohrimenko1988 I fixed that already. Please, review first report since others are related to minor changes that do not influence functionality

riuvshin commented:
Results of automated E2E tests of Eclipse Che Multiuser on OCP:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1207/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1207//Selenium_20tests_20report/)
docker image: eclipseche/che-server:11890
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

garagatyi commented:
@Ohrimenko1988 tests finished. Could you take a look?

Ohrimenko1988 commented:
Tests result look really perfect. From QE side no objection.

garagatyi commented:
@Ohrimenko1988 thanks!
